core/src/main/java/io/grpc/Metadata.java

```diff
@@ -177,6 +178,14 @@ public T apply(MetadataEntry entry) {
   }
 
   /**
+   * Returns set of all keys in store
+   * @return ImmutableSet of keys
+   */
+  public Set<String> keys() {
+    return ImmutableSet.copyOf(store.keySet());
```

Comment:
If you don't plan on changing it after getting the keys, you could probably just use Collections.unmodifiableSet, which avoids the copy.